src/channel.cr

```diff
@@ -203,6 +203,33 @@ abstract class Channel(T)
       @channel.unwait_for_send
     end
   end
+
+  # :nodoc:
+  struct TimeoutAction
+    include SelectAction
+
+    @timeout_at : Time::Span
+
+    def initialize(timeout : Time::Span)
```

Comment:
I'd prefer to do the `timeout_at` calculation in `timeout_select_action`. That means a `timeout_at_select_action` can be added in the future, which accepts a timestamp instead of a time from now.